File: tools/denovate-opentelemetry-from-alloy.py
#!/usr/bin/env python3

import argparse
import json
import subprocess
import sys
import tempfile
from pathlib import Path

OTEL_PREFIX = "go.opentelemetry.io/"


def run(cmd: list[str], cwd: Path | None = None) -> str:
    result = subprocess.run(cmd, cwd=cwd, text=True, capture_output=True)
    if result.returncode != 0:
        if result.stdout:
            print(result.stdout, file=sys.stderr, end="")
        if result.stderr:
            print(result.stderr, file=sys.stderr, end="")
        raise SystemExit(result.returncode)
    return result.stdout.strip()


def read_otel_requirements(go_mod_path: Path) -> dict[str, str]:
    raw = run(["go", "mod", "edit", "-json", str(go_mod_path)])
    data = json.loads(raw)
    requirements: dict[str, str] = {}
    for req in (data.get("Require") or []):
        dep_path = req["Path"]
        if dep_path.startswith(OTEL_PREFIX):
            requirements[dep_path] = req["Version"]
    return requirements


def collect_alloy_versions(alloy_dir: Path) -> dict[str, str]:
    go_mod_files = [
        line.strip()
        for line in run(["git", "-C", str(alloy_dir), "ls-files", "**/go.mod"]).splitlines()
        if line.strip()
    ]

    versions: dict[str, str] = {}
    sources: dict[str, str] = {}
    for rel_path in sorted(go_mod_files):
        go_mod = alloy_dir / rel_path
        for dep_path, version in read_otel_requirements(go_mod).items():
            current = versions.get(dep_path)
            if current is not None and current != version:
                src = sources[dep_path]
                raise SystemExit(
                    "Failed: conflicting versions in alloy for "
                    f"{dep_path}: {current} (from {src}) vs {version} (from {rel_path})"
                )
            versions[dep_path] = version
            sources[dep_path] = rel_path
    return versions


def apply_versions(repo_root: Path, versions: dict[str, str], deps: list[str]) -> None:
    for dep in deps:
        version = versions.get(dep)
        if version is None:
            raise SystemExit(f"Failed: dependency {dep} was not found in alloy")
        print(f"  - {dep} => {version}")
        run(["go", "mod", "edit", f"-require={dep}@{version}"], cwd=repo_root)


def verify_versions(
    go_mod_path: Path,
    expected_versions: dict[str, str],
    original_deps: list[str],
) -> None:
    final_versions = read_otel_requirements(go_mod_path)
    for dep in original_deps:
        expected = expected_versions.get(dep)
        if expected is None:
            raise SystemExit(f"Failed: no expected alloy version found for {dep}")

        final = final_versions.get(dep)
        if final is None:
            raise SystemExit(
                f"Failed: {dep} disappeared after go mod tidy (expected {expected})"
            )
        if final != expected:
            raise SystemExit(
                f"Failed: {dep} is {final} after go mod tidy, expected {expected}"
            )


def main() -> int:
    parser = argparse.ArgumentParser(
        description=(
            "Sync go.opentelemetry.io/* dependencies in this repo to versions "
            "used by a specified grafana/alloy revision."
        )
    )
    parser.add_argument("alloy_revision", help="Alloy tag, branch, or commit")
    args = parser.parse_args()

    script_dir = Path(__file__).resolve().parent
    repo_root = script_dir.parent
    go_mod_path = repo_root / "go.mod"
    if not go_mod_path.exists():
        raise SystemExit(f"Failed: go.mod not found at {go_mod_path}")

    profiler_versions = read_otel_requirements(go_mod_path)
    if not profiler_versions:
        print("No go.opentelemetry.io/* dependencies found; nothing to do.")
        return 0

    original_deps = sorted(profiler_versions.keys())

    with tempfile.TemporaryDirectory() as tmp:
        alloy_dir = Path(tmp) / "alloy"
        print("Cloning grafana/alloy...")
        run(["git", "clone", "https://github.com/grafana/alloy.git", str(alloy_dir)])
        run(["git", "-C", str(alloy_dir), "checkout", args.alloy_revision])
        resolved_revision = run(["git", "-C", str(alloy_dir), "rev-parse", "HEAD"])

        print(
            "Collecting go.opentelemetry.io/* dependency versions from alloy "
            f"({resolved_revision})..."
        )
        alloy_versions = collect_alloy_versions(alloy_dir)

    print("Applying alloy versions to profiler go.opentelemetry.io/* dependencies...")
    apply_versions(repo_root, alloy_versions, original_deps)

    print("Running go mod tidy...")
    run(["go", "mod", "tidy"], cwd=repo_root)

    print("Verifying resulting versions...")
    verify_versions(go_mod_path, alloy_versions, original_deps)

    print(
        "Success: all go.opentelemetry.io/* dependencies match alloy revision "
        f"{args.alloy_revision}."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/denovate-opentelemetry-from-alloy.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"

	"golang.org/x/mod/modfile"
)

const (
	otelPrefix  = "go.opentelemetry.io/"
	alloyModule = "github.com/grafana/alloy"
)

var skippedDeps = map[string]struct{}{
	"go.opentelemetry.io/proto/otlp":                        {},
	"go.opentelemetry.io/proto/otlp/profiles/v1development": {},
}

func shouldSkipDep(dep string) bool {
	_, ok := skippedDeps[dep]
	return ok
}

type goModDownloadResult struct {
	Dir     string `json:"Dir"`
	Version string `json:"Version"`
	Error   string `json:"Error"`
}

func run(repoRoot string, name string, args ...string) (string, error) {
	cmd := exec.Command(name, args...)
	cmd.Dir = repoRoot

	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		return "", fmt.Errorf(
			"%s %s failed: %w\nstdout:\n%s\nstderr:\n%s",
			name,
			strings.Join(args, " "),
			err,
			stdout.String(),
			stderr.String(),
		)
	}
	return strings.TrimSpace(stdout.String()), nil
}

func readModFile(path string) (*modfile.File, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", path, err)
	}

	parsed, err := modfile.Parse(path, content, nil)
	if err != nil {
		return nil, fmt.Errorf("parse %s: %w", path, err)
	}
	return parsed, nil
}

func otelRequirements(parsed *modfile.File) map[string]string {
	reqs := make(map[string]string)
	for _, req := range parsed.Require {
		if strings.HasPrefix(req.Mod.Path, otelPrefix) {
			reqs[req.Mod.Path] = req.Mod.Version
		}
	}
	return reqs
}

func sortedKeys(m map[string]string) []string {
	keys := make([]string, 0, len(m))
	for key := range m {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}

func downloadAlloy(repoRoot string, revision string) (dir string, resolvedVersion string, err error) {
	out, err := run(repoRoot, "go", "mod", "download", "-json", alloyModule+"@"+revision)
	if err != nil {
		return "", "", err
	}

	var result goModDownloadResult
	if unmarshalErr := json.Unmarshal([]byte(out), &result); unmarshalErr != nil {
		return "", "", fmt.Errorf("decode go mod download output: %w\nraw output:\n%s", unmarshalErr, out)
	}
	if result.Error != "" {
		return "", "", errors.New(result.Error)
	}
	if result.Dir == "" {
		return "", "", fmt.Errorf("go mod download did not return Dir for %s", revision)
	}
	if result.Version == "" {
		result.Version = revision
	}
	return result.Dir, result.Version, nil
}

func applyAlloyVersions(repoRoot string, profilerDeps map[string]string, alloyDeps map[string]string) error {
	for _, dep := range sortedKeys(profilerDeps) {
		currentVersion := profilerDeps[dep]
		if shouldSkipDep(dep) {
			fmt.Printf("  - %s: %s => %s (skipping dependency)\n", dep, currentVersion, currentVersion)
			continue
		}

		alloyVersion, ok := alloyDeps[dep]
		if !ok {
			return fmt.Errorf("dependency %s was not found in alloy", dep)
		}
		fmt.Printf("  - %s: %s => %s\n", dep, currentVersion, alloyVersion)
		if _, err := run(repoRoot, "go", "mod", "edit", "-require="+dep+"@"+alloyVersion); err != nil {
			return err
		}
	}
	return nil
}

func verifyAligned(profilePath string, profilerDeps map[string]string, alloyDeps map[string]string) error {
	finalMod, err := readModFile(profilePath)
	if err != nil {
		return err
	}
	finalDeps := otelRequirements(finalMod)

	for _, dep := range sortedKeys(profilerDeps) {
		if shouldSkipDep(dep) {
			continue
		}

		expected, ok := alloyDeps[dep]
		if !ok {
			return fmt.Errorf("no expected alloy version found for %s", dep)
		}

		finalVersion, ok := finalDeps[dep]
		if !ok {
			return fmt.Errorf("%s disappeared after go mod tidy (expected %s)", dep, expected)
		}
		if finalVersion != expected {
			return fmt.Errorf("%s is %s after go mod tidy, expected %s", dep, finalVersion, expected)
		}
	}
	return nil
}

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <alloy-revision>\n", os.Args[0])
		os.Exit(1)
	}
	alloyRevision := os.Args[1]

	repoRoot, err := os.Getwd()
	if err != nil {
		fmt.Fprintf(os.Stderr, "failed to get cwd: %v\n", err)
		os.Exit(1)
	}
	goModPath := filepath.Join(repoRoot, "go.mod")

	profilerMod, err := readModFile(goModPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "failed to read profiler go.mod: %v\n", err)
		os.Exit(1)
	}
	profilerDeps := otelRequirements(profilerMod)
	if len(profilerDeps) == 0 {
		fmt.Println("No go.opentelemetry.io/* dependencies found; nothing to do.")
		return
	}

	fmt.Println("Downloading grafana/alloy with go mod...")
	alloyDir, resolvedRevision, err := downloadAlloy(repoRoot, alloyRevision)
	if err != nil {
		fmt.Fprintf(os.Stderr, "failed to download alloy module: %v\n", err)
		os.Exit(1)
	}

	alloyGoModPath := filepath.Join(alloyDir, "go.mod")
	alloyMod, err := readModFile(alloyGoModPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "failed to read alloy go.mod: %v\n", err)
		os.Exit(1)
	}
	alloyDeps := otelRequirements(alloyMod)

	fmt.Printf("Collecting go.opentelemetry.io/* dependency versions from alloy go.mod (%s)...\n", resolvedRevision)
	fmt.Println("Applying alloy versions to profiler go.opentelemetry.io/* dependencies...")
	if err := applyAlloyVersions(repoRoot, profilerDeps, alloyDeps); err != nil {
		fmt.Fprintf(os.Stderr, "failed to apply alloy versions: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Running go mod tidy...")
	if _, err := run(repoRoot, "go", "mod", "tidy"); err != nil {
		fmt.Fprintf(os.Stderr, "go mod tidy failed: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Verifying resulting versions...")
	if err := verifyAligned(goModPath, profilerDeps, alloyDeps); err != nil {
		fmt.Fprintf(os.Stderr, "verification failed: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Success: all go.opentelemetry.io/* dependencies match alloy revision %s.\n", alloyRevision)
}
